### PR TITLE
Lrdocs 3033 adaptive media

### PR DESCRIPTION
**Pull request by mwilliams2014**
@sergiogonzalez I've added some questions that I have on the Adaptive Media APIs in this pr to myself. I tried sending one to you, but the changes were to big and caused the page to lag :-)

Please review and let me know what you think. Thanks!

I've also added the userguide docs draft to this pull as well and added a question. Please review and let me know what you think. thx!

### CODE REVIEW

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown [hunk 1]
```diff
@@ -0,0 +1,283 @@
+# Configuring Adaptive Media APIs
+
+The Adaptive Media APIs give you lots of control over your media. You can create
+unique attributes for your media types, search and obtain information about
+adaptive media such as the width and height of an image, and obtain URLs for an 
+adaptive media. This tutorial demonstrates these features.
+
+<!-- Are there any other things you can do with the API that we should mention? -->
```
Comment (mwilliams2014) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: Are there any other things you can do with the API that we should mention?

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+First, you must deploy the proper Adaptive Media OSGi modules. This is covered
+next.
+
+## Deploying Adaptive Media
+
+Development of Adaptive Media modules is done in the master branch of an
+independent repository of Liferay at the following URL:
+[https://github.com/liferay/com-liferay-adaptive-media](https://www.google.com/url?q=https://github.com/liferay/com-liferay-adaptive-media&sa=D&ust=1482880889869000&usg=AFQjCNHzWR-uBL7ROpGnHzG2agFqs89miQ)
+
+<!-- Shouldn't we just point to the Liferay-portal repo and consider that what
```
Comment (mwilliams2014) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: Shouldn't we just point to the Liferay-portal repo and consider that what is in Portal is always stable? I know at the moment the repo isn't up to date in Portal.

I still mention the other repo if they want the bleeding edge changes.

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+-   A set of attributes (content length, content type, dimensions,
+    etc.).
+-   The raw content of the media.
+-   A URI that uniquely identifies the media.
+
+The set of attributes allowed for a media depend on its type. For example, 
+images have different characteristics from videos or audio files, and the API 
+tries to enforce this statically.
+
+<!-- How do you create a media type? Should we mention that in this tutorial, or
```
Comment (mwilliams2014) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: How do you create a media type? Should we mention that in this tutorial, or 
just focus this one on Adaptive Media Images?

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            queryBuilder.allForFileEntry(fileEntry).
+    
+            orderBy(ImageAdaptiveMediaAttribute.IMAGE_WIDTH, true).
+    
+            done());
+
+This returns a stream of the image variants created for the image fileEntry 
+in ascending order by width.
+
+<!-- Is this using an InputStream to get the content or is this completely
```
Comment (mwilliams2014) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: Is this using an InputStream to get the content or is this completely different?

    Stream<AdaptiveMedia<ImageAdaptiveMediaProcessor>> stream =
    
      _finder.getAdaptiveMedia(
    
         queryBuilder ->
    
            queryBuilder.
    
               forVersion(_fileVersion).
    
               with(ImageAdaptiveMediaAttribute.IMAGE_HEIGHT, 100).
    
               done());

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+`AdaptiveMediaAttributeConverterUtil::parseInt` as the converter function, and 
+`ImageAdaptiveMediaAttribute::_intDistance` as the comparator.
+Finally, the new `height` and `width` attributes for the 
+`ImageAdaptiveMediaAttribute` are added to the `_allowedAttributes` HashMap.
+
+The contents of a media are exposed via an InputStream. Once requested
+with the [`getInputStream()` method](https://github.com/liferay/com-liferay-adaptive-media/blob/6356d57a76050764d52a8dccce07578cc53b72b1/adaptive-media-api/src/main/java/com/liferay/adaptive/media/AdaptiveMedia.java#L45-L52), 
+consumers are responsible for closing it when finished.
+
+<!-- Is there an example of using and closing an InputStream? Is this
```
Comment (mwilliams2014) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: Is there an example of using and closing an InputStream? 

Is this ImageProcessor class an example of one? [https://github.com/liferay/com-liferay-adaptive-media/blob/5220d012bc037a1736746eddc9a0ab9aa90ea8d4/adaptive-media-image-impl/src/main/java/com/liferay/adaptive/media/image/internal/util/ImageProcessor.java](https://github.com/liferay/com-liferay-adaptive-media/blob/5220d012bc037a1736746eddc9a0ab9aa90ea8d4/adaptive-media-image-impl/src/main/java/com/liferay/adaptive/media/image/internal/util/ImageProcessor.java)-->

File: discover/portal/articles/03-collaboration/01-managing-documents-and-media/08-adapting-your-media.markdown [hunk 6]
```diff
@@ -1,308 +1,165 @@
...
+-   Choose the image whose height is smaller than the screen height.
+-   Choose the image whose width and height are smaller than the
+    screen width and height.
+-   Choose the image whose width and height are more similar to the
+    screen size.
+-   Choose the image whose width and height are slightly smaller than
+    the screen size
 
-  ImageAdaptiveMediaAttribute.IMAGE\_HEIGHT)
+As you can see, there are many ways to choose the best image for a device. The
```
Comment (mwilliams2014) [inline discover/portal/articles/03-collaboration/01-managing-documents-and-media/08-adapting-your-media.markdown]: > The consumer of the Adaptive Media API will decide which image is best for the 
particular use case.

@sergiogonzalez How does the consumer specify which image variant to use with the API? Is that something you would want to do, or is this automatically determined?

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown (hunk 1 above)
Comment (sergiogonzalez) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: I think that's mainly it.

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown (hunk 2 above)
Comment (sergiogonzalez) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: Hey @mwilliams2014 yeah, you're totally right. The liferay-portal repository should contain the latest stable version of ADaptive Meida while the independent repo might contain features not ready to release yet.

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown (hunk 3 above)
Comment (sergiogonzalez) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: I think that we should focus on Adaptive Media images for now. In the future we might expand it, but let's focus on images for now.

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown (hunk 5 above)
Comment (sergiogonzalez) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: The input stream is just a regular java object with a well defined API so I don't think that we need to add any more explanation about it: https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html

In order to close you just need to do inputStream.close()

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown (hunk 4 above)
Comment (sergiogonzalez) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: InputStream (https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html) and Stream (https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html) are 2 completely different things and they are not related at all.

In this case we are talking about Stream and they are not InputStream.

File: discover/portal/articles/03-collaboration/01-managing-documents-and-media/08-adapting-your-media.markdown (hunk 6 above)
Comment (sergiogonzalez) [inline discover/portal/articles/03-collaboration/01-managing-documents-and-media/08-adapting-your-media.markdown]: Hey @mwilliams2014 the consumer can specify what the query provides. Take a look at the tests here https://github.com/liferay/com-liferay-adaptive-media/blob/master/adaptive-media-image-impl/src/test/java/com/liferay/adaptive/media/image/internal/finder/ImageAdaptiveMediaQueryBuilderImplTest.java#L72

You can see that the query is implemented using a Builder pattern that "tells" the developer the different options available.

**Comment by sergiogonzalez** [comment]
Hey @mwilliams2014 I think I replied to all your comments. 
Let me know if there's anything I'm missing.

Thanks!

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown (hunk 5 above)
Comment (mwilliams2014) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: Ok. Yeah, we don't need to explain that one then. thx!

File: develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown (hunk 4 above)
Comment (mwilliams2014) [inline develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown]: Thanks for clarifying!

File: discover/portal/articles/03-collaboration/01-managing-documents-and-media/08-adapting-your-media.markdown (hunk 6 above)
Comment (mwilliams2014) [inline discover/portal/articles/03-collaboration/01-managing-documents-and-media/08-adapting-your-media.markdown]: thx!

**Pull request closed without merging**